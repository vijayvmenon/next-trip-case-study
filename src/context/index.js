import { useContext } from 'react';
import AppContext, { Context } from './context';

export default AppContext;
export { Context };
export const useMyContext = () => useContext(Context);
