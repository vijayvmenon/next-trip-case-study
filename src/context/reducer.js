import constants from '../constants';
import ENTranslations from '../locales/en.json';
import { localStorageExists } from '../utils';

const localStorageLanguage =
  localStorageExists && localStorage.getItem('LANGUAGE');
const initialState = {
  tableExpanded: false,
  hideHeader: false,
  snackbar: { value: '', text: '' },
  language: localStorageLanguage || 'en',
  translations: localStorageLanguage
    ? constants.LANGUAGE_MAPPING[localStorageLanguage]
    : ENTranslations,
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case 'TABLE_EXPAND':
      return { ...state, tableExpanded: true };
    case 'TABLE_COLLAPSE':
      return { ...state, tableExpanded: false };
    case 'SET_SNACKBAR':
      return { ...state, snackbar: { value: action.value, text: action.text } };
    case 'HIDE_HEADER':
      return { ...state, hideHeader: true };
    case 'SHOW_HEADER':
      return { ...state, hideHeader: false };
    case 'SET_LANGUAGE':
      return {
        ...state,
        language: action.value,
        translations: constants.LANGUAGE_MAPPING[action.value],
      };
    default:
      return { ...state };
  }
}

export { initialState };
