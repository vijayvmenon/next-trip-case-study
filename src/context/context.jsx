import React, { useReducer, createContext } from 'react';
import PropTypes from 'prop-types';
import reducer, { initialState } from './reducer';

export const Context = createContext();

export default function AppContext({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  );
}

AppContext.propTypes = {
  children: PropTypes.objectOf(PropTypes.any).isRequired,
};
