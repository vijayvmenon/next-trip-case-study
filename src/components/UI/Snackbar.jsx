import React, { useEffect, useContext } from 'react';
import { createUseStyles } from 'react-jss';
import { useLocation } from 'wouter';
import constants, { colors } from '../../constants';
import { ReactComponent as CloseIcon } from '../../assets/close-icon-snackbar.svg';
import { Context } from '../../context';

const useStyles = createUseStyles({
  hidden: {
    display: 'none',
  },
  root: {
    display: 'flex',
    alignItems: 'center',
    margin: 8,
    position: 'fixed',
    right: 0,
    left: 0,
    zIndex: 1,
    transition: 'opacity .15s,transform .15s',
    pointerEvents: 'none',
    justifyContent: 'center',
    bottom: 8,
  },
  wrapper: {
    color: 'rgb(255, 255, 255)',
    fontSize: '1rem',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-start',
    borderRadius: 4,
    minWidth: 334,
    maxWidth: 672,
    boxShadow:
      '0 3px 5px -1px rgb(0 0 0 / 20%), 0 6px 10px 0 rgb(0 0 0 / 14%), 0 1px 18px 0 rgb(0 0 0 / 12%)',
    fontWeight: 400,
    letterSpacing: '1px',
    lineHeight: '20px',
    textAlign: 'left',
  },
  error: {
    extend: 'wrapper',
    backgroundColor: 'rgb(244, 67, 54)',
  },

  warning: { extend: 'wrapper', backgroundColor: 'rgb(255, 152, 0)' },
  snackbarText: {
    flexGrow: 1,
    padding: '14px 16px',
    margin: 0,
    pointerEvents: 'auto',
  },
  button: {
    color: 'rgb(255, 255, 255)',
    fontSize: '1rem',
    flexShrink: 0,
    boxSizing: 'border-box',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: 36,
    height: 36,
    padding: 8,
    margin: '0 8px 0 0',
    cursor: 'pointer',
    position: 'relative',
    pointerEvents: 'auto',
    WebkitTapHighlightColor: 'transparent',
    outline: 'none',
    backgroundColor: 'transparent',
    border: 'none',
  },
  icon: {
    width: '1rem',
    height: '1rem',
    color: colors.WHITE,
  },
});

export default function Snackbar() {
  const classes = useStyles();
  const {
    dispatch,
    state: { snackbar: { value: snackbarType = '', text = '' } = {} } = {},
  } = useContext(Context);
  const [location] = useLocation();
  useEffect(() => {
    if (snackbarType) {
      const timeout = setTimeout(() => {
        dispatch({ type: 'SET_SNACKBAR', value: '', text: '' });
      }, constants.SNACKBAR_TIMEOUT);

      return () => clearTimeout(timeout);
    }
    return undefined;
  }, [dispatch, snackbarType]);

  // Reset snackbar on location change, i.e on any redirection to another url within app
  useEffect(() => {
    dispatch({ type: 'SET_SNACKBAR', value: '', text: '' });
  }, [dispatch, location]);

  return (
    snackbarType && (
      <div className={classes.root}>
        <div
          data-testid="custom-snackbar"
          className={snackbarType === 'error' ? classes.error : classes.warning}
        >
          <div className={classes.snackbarText}>{text}</div>
          <button
            type="button"
            onClick={() => dispatch({ type: 'SET_SNACKBAR', value: '' })}
            className={classes.button}
          >
            <CloseIcon className={classes.icon} />
          </button>
        </div>
      </div>
    )
  );
}
