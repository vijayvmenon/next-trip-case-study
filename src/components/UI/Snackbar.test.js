import { render, fireEvent, screen } from '@testing-library/react';
import Snackbar from './Snackbar';
import renderer from 'react-test-renderer';
import { Context } from '../../context';

const dispatch = jest.fn(); // mock dispatch

const props = {
  hrefLink: '/test',
  name: 'test',
};

describe('Component:Snackbar.jsx', () => {
  test('Snackbar.jsx renders correctly', () => {
    const snackbar = renderer
      .create(
        <Context.Provider value={{ dispatch }}>
          <Snackbar {...props} />
        </Context.Provider>,
      )
      .toJSON();
    expect(snackbar).toMatchSnapshot();
  });

  test('Snackbar is displayed only when snackbar value is set from context', () => {
    const { container, getByTestId } = render(
      <Context.Provider
        value={{ dispatch, state: { snackbar: { value: 'error' } } }}
      >
        <Snackbar {...props} />
      </Context.Provider>,
    );
    expect(getByTestId('custom-snackbar')).toBeTruthy();
  });
});
