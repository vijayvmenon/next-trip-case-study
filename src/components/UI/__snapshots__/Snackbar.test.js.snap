// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:Snackbar.jsx Snackbar.jsx renders correctly 1`] = `""`;
